### ISSUE TITLE
Some files are randomly being missed

### ISSUE THREAD

**Issue by darkalor** [whitecolor/ts-node-dev#181]
Platform: Window 10
ts-node-dev version: 1.0.0-pre.44, also tried with 1.0.0-pre.52

Running into this really bizarre behavior - at first start tsnd can miss some files that have to be compiled. After doing a couple of restarts (without any changes, e.g. just by re-saving a file) it manages to pick up everything and works until the next time I re-run it.

Here's one complete fail at first run, doesn't pick up any files past `index`
![image](https://user-images.githubusercontent.com/2834813/87931901-3bc52300-ca93-11ea-862f-4ff03c8859fb.png)

After a restart - a couple more files got picked up
![image](https://user-images.githubusercontent.com/2834813/87931929-47b0e500-ca93-11ea-85f7-38e9dcb32aa3.png)

One more restart - the module `Schema` only picks up two files, but should contain the same files as the module `Event`
![image](https://user-images.githubusercontent.com/2834813/87932100-9fe7e700-ca93-11ea-846a-810b0197c844.png)

**Comment by whitecolor** [comment]
Check this https://github.com/whitecolor/ts-node-dev/issues/169#issuecomment-657187270

**Comment by darkalor** [comment]
Doesn't seem like it's the same issue. I don't have a problem with type files being picked up - its the actual sources that get missed.  

I've also been unable to reproduce this when just running with `ts-node`. The issue happens only with `ts-node-dev`

**Comment by DarkLite1** [comment]
I experienced the [same issue](https://github.com/whitecolor/ts-node-dev/issues/162) but recently after upgrade to `"ts-node-dev": "^1.0.0-pre.55` and using the latest TypeScript `3.9.7` this seems to be gone.

**Comment by darkalor** [comment]
I didn't realize that issue would be related, since it mentions issues appearing after `1.0.0-pre.44`, but I already had them on 44.
Updating to 55 TS 3.9.7 did not resolve the issue for me.
It seems to work *slightly* better though - it seems to require only one restart now. Previously I had to do it 2-3 times sometimes.

**Comment by DarkLite1** [comment]
Yeah, I've been in that boat too. Restarting for over 5 times sometimes before all files were correctly compiled. Usually I did this by hitting `CTRL + S` a couple of times on the ``index.ts`  file. 

This is indeed a very annoying issue... It gets better with every release but I feel there should be a test suite that verifies if all files are completely compiled when `ts-node-dev` is done. Part of the issue is also that it doesn't happen for everyone. It's like on one OS/machine it does and the next one is fine, see my previous issue. 

This is a hard one to tackel. Maybe @whitecolor could add some better/more detailed logging when running with the `--debug` argument. It would allow for more detailed and accurate reports from users when things go wrong.

**Comment by DarkLite1** [comment]
> I've also been unable to reproduce this when just running with `ts-node`. The issue happens only with `ts-node-dev`

Exactly

**Comment by whitecolor** [comment]
`--debug` outputs all files being required and compiled, are your missing files not in this log, they are not being required?

**Comment by whitecolor** [comment]
@darkalor can you show the code of your `index.ts` from example above.

**Comment by darkalor** [comment]
>  --debug outputs all files being required and compiled, are your missing files not in this log, they are not being required?

It looks like files are not getting required, even though they're imported in the index. When I just restart ts-node-dev a couple of times on the exact same code it picks up more files in the list each time (see my original screenshots). It looks like it's not waiting for some async stuff perhaps, and just skips forward sometimes without getting everything compiled. 

The behavior changed a bit with .55 - I'm getting a failure more consistently in the exact same spot now.
I've also noticed that if I reduce the amount of imports it fails less often.

Here's how the index looks:

```
import dotenv from 'dotenv'
dotenv.config({ path: '../.env' })
import consoleStamp from 'console-stamp'
consoleStamp(console)
import express from 'express'
import { ApolloServer } from 'apollo-server-express'
import graphqlPlayground from 'graphql-playground-middleware-express'
import mongoose from 'mongoose'
import { typeDefs, resolvers } from './modules'
import * as auth from './auth'

async function main() {
  try {
    // Express
    const app = express()

    // Authentication
    console.log(auth.authenticate)            //<----------- this is where undefined prints in the error sceenshot
    app.use('/graphql', auth.authenticate)

    // GraphQL Playground
    app.get('/playground', graphqlPlayground({ endpoint: '/graphql' }))

    // MongoDB
    const db = await mongoose.connect(process.env.MONGO_CONNECTION_STRING, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    })
    db && console.log('🍃 Connected to MongoDB')

    // TODO: trying to catch issue with builds failing
    console.log(typeDefs)                        // <-------- this one sometimes gives a bad result too

    // Apollo Server
    const server = new ApolloServer({
      playground: false,
      introspection: true,
      typeDefs,
      resolvers,
    })
    server.applyMiddleware({ app })

    // Start Apollo Server
    app.listen({ port: 8081 }, async () => {
      console.log(
        `🚀 Server ready at http://localhost:8081/playground`
      )
    })
  } catch (err) {
    console.error(err)
  }
}

main()
```

And that's the error I get. It even says that it compiled `auth.js` but the console log right after that prints it as `undefined`
![image](https://user-images.githubusercontent.com/2834813/88381334-e64c8700-cdae-11ea-871d-4be703f45894.png)

**Comment by DarkLite1** [comment]
@darkalor  Just one stupid question here: are you certain that you don't need to do `await auth.authenticate`? That is't not a promise? Just to be sure...

**Comment by darkalor** [comment]
@DarkLite1 Nope it's not a promise. Besides if it was - the log would have printed that it's a promise.

**Comment by whitecolor** [comment]
@darkalor Is something wrong with deps required by `auth` module? Maybe something wrong with its compilation?

**Comment by samelie** [comment]
for what it's worth, I came across this issue because I'm experiencing something similar where it will print eg: ```
[PEPPER] [INFO] 22:49:29 Restarting: src/redis/index.ts has been modified
[PEPPER] [DEBUG] 22:49:29 Removing all watchers from files
[PEPPER] [DEBUG] 22:49:29 Child is still running, restart upon exit
[PEPPER] [DEBUG] 22:49:29 Disconnecting from child
[PEPPER] [DEBUG] 22:49:29 Sending SIGTERM kill to child pid 52499
[PEPPER] [DEBUG] 22:49:29 Using tree-kill
[PEPPER] Child got SIGTERM, exiting.
[PEPPER] [DEBUG] 22:49:29 Child exited with code 0
```
 But then the change will **not** be reflected - was trying with `console.logs` , using `tsnd  --exit-child --tree-kill  --respawn  --debug  --watch  'src/**/*.ts'  src/index.ts` . 
- First I add a log, --> see the log
- Then I remove it , <does the change, debug log seems okay> , _the console log is still there_

What is interesting is when I change the `return` value of a function say, the change is picked up, tho the console.log _is still there_

I'm running some more tests around `const a = 2; const b = 3; const c = a + b;`

**Comment by whitecolor** [comment]
Maybe reproducable  example?

**Comment by darkalor** [comment]
> 
> 
> @darkalor Is something wrong with deps required by `auth` module? Maybe something wrong with its compilation?

Apologies for the delayed response. No it's not related to the `auth` module. Compilation can fail before or after that randomly.

**Comment by DarkLite1** [comment]
@darkalor I had the same problem for reproducing an example that demonstrates the problem. @samelie you seem to have a more simple examples, would you mind posting it here for further testing? Or a link to your GitHub?

**Comment by GV94** [comment]
I encountered this issue exactly as you describe it today. It was working fine for the longest time, and then it suddenly stopped being able to resolve all the sources imported in the index file (there were no changes to the code or dependencies). Saving a couple of times to trigger a restart solves the problem and the code runs eventually. We are also using `tsconfig-paths`, and it works for every member of the team except for me.

Running the code with ts-node works, as does nodemon.
